add getting started doc

minrk opened the pull request:
Covering step by step approach to common configuration. It's not done, but is at least complete enough for feedback. Any important ideas that I missed, or anything not make sense?

/cc @ellisonbg @jhamrick

I don't want to get *too much* detail into this particular doc, but it would be great to get some feedback on the balance between brevity / approachability and detailed coverage.

ellisonbg commented:
Great I will have a look at this!

ssanderson commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+
+## external services
+
+JupyterHub has a REST API that can be used
```
While technically true, I'm guessing more words were planned for this sentence?

minrk replied:
ha, yes. I might remove the external services bit from this doc. I was originally thinking of documenting using `jupyterhub token` and setting up the culling script, but I might save that for a more dedicated doc.

minrk commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+
+# specify users and admin
+c.Authenticator.whitelist = {'rgbkrk', 'minrk', 'jhamrick'}
+c.JupyterHub.admin_users = {'jhamrick', 'rgbkrk'}
```
This block tells me that I should probably put the admin user list on the Authenticator, even though no Authenticator methods actually check admin.

jhamrick replied:
Yeah, that would certainly make it more consistent. Maybe the authenticator could check what group a user is in after authenticating (currently it would just be "admin" or nothing, but maybe once there is collaboration stuff there could be more groups), and then JupyterHub queries the authenticator for that information?

minrk replied:
moved to Authenticator

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+If you are going to separate these processes across machines or containers,
+you may need to tell them to listen on addresses other than localhost.
+
+**NOTE** this server is running without SSL encryption.
```
I might move this note further up, directly before the "authentication" section -- otherwise people might miss it, and I think it's an important thing to make sure they know.

jhamrick replied:
You could modify this note to mention that configuring JupyterHub with SSL encryption is mentioned later, and then link to that section.

ellisonbg replied:
+1

minrk replied:
done

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+  This file allows the Hub to remember what users are running and where,
+  as well as other information enabling you to 
+  You can change the location of this file with `--db=/path/to/somedb.sqlite`.
+- `jupyterhub_cookie_secret` is the encryption key used for securing cookies.
```
Does it get overwritten by default if the server is restarted, or will the hub use it if it exists? I thought that in order to make sure it was the same you had to pass it in as `JPY_COOKIE_SECRET`? Or is that just an alternative way to do it?

jhamrick replied:
Ah, I see, you discuss this more later down below. In that case, I'd just add a quick note here saying that the cookie secret is discussed in more detail further down, and then link to it.

minrk replied:
link added

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+  Conversely, deleting this file and restarting the server effectively invalidates all login cookies.
+
+
+## How to configure JupyterHub
```
It might be useful to provide a link here to how IPython/Jupyter config files work in general.

minrk replied:
link added

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+- the Hub itself
+
+The proxy is the public-facing part of the application.
+The default public IP is `''`, which means all interfaces on the machine.
```
Maybe say "The default public IP is `''` (equivalent to `0.0.0.0`), which means all interfaces on the machine."

minrk replied:
done

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+The default port is 8000.
+If you want to specify where the Hub application as a whole can be found,
+modify these two values.
+If you want to listen on a particular IP,
```
*For example*, if you want to listen...

minrk replied:
done

jhamrick commented on docs/getting-started.md:
````diff
@@ -0,0 +1,325 @@
+```python
+# ideally a private network address
+c.JupyterHub.proxy_api_ip = '10.0.1.4'
+c.JupyterHub.proxy_api_port = 5432
````
Maybe mention what the default port is too?

minrk replied:
done, for both proxy and hub

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+then admin users have permission to log in *as other users* on their respective machines,
+for debugging. **You should make sure your users know if admin_access is enabled.**
+
+### adding and removing users
```
"adding" should be capitalized

minrk replied:
done

jhamrick commented on docs/getting-started.md:
````diff
@@ -0,0 +1,325 @@
+c.LocalAuthenticator.create_system_users = True
+```
+
+however, adding a user to the Hub that doesn't already exist on the system will result
````
I'm not sure if this is really the right place for it, but I feel like it would be good to mention somewhere that if you add users via JupyterHub and then restart JupyterHub, they will still be on the whitelist as long as the database still exists. If the database is removed, then JupyterHub will start over from the whitelist you give it, meaning you either need to adjust that whitelist to include the users you added, or re-add them once JupyterHub has started.

minrk replied:
added at the top of this section, since it is generic to adding users no matter what the Authenticator.

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+which is the place to put configuration that you want to affect all of your users.
+
+
+- setting working directory
```
Are these remnants from the outline?

minrk replied:
yup, removed

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+
+You are hosting JupyterHub on a single cloud server,
+using https on the standard https port, 443.
+You want to use GitHub OAuth for login,
```
I would add a link here to the GitHub OAuth repository.

minrk replied:
done

jhamrick commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+# with Welcome.ipynb as the default landing page
+# this config could also be put in
+# /etc/ipython/ipython_notebook_config.py
+c.Spawner.notebook_dir = '~/assignments'
```
Should this be `~/notebooks`?

minrk replied:
I switched it to assigments to avoid confusion, since the URLs start with `/notebooks/` regardless. I updated the goals described at the top, so it's consistent.

jhamrick commented on docs/getting-started.md:
````diff
@@ -0,0 +1,325 @@
+c.Spawner.args = ['--NotebookApp.default_url=/notebooks/Welcome.ipynb']
+```
+
+Using the GitHub Authenticator requires a few env variables,
````
I'd add a link here to the oauthenticator documentation explaining how they get these variables.

Also maybe say "a few *additional* env variables", as a reminder that they can also set things like `CONFIGPROXY_AUTH_TOKEN`.

minrk replied:
link added, CONFIGPROXY_AUTH_TOKEN added to the example, as well.

jhamrick commented:
This looks great, Min! I left some comments inline, but mostly minor stuff. I would maybe just add a few more pointers at the end for next steps:

* if it's not working, try getting vanilla jupyterhub working first and add in one component at a time
* links to the wiki pages that have various existing authenticators and spawners

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+
+The default Authenticator that ships with JupyterHub
+authenticates users with their system name and password (via [PAM][]).
+Any user on the system with a password will be allowed to start a notebook server.
```
start a single user notebook server

minrk replied:
:heavy_check_mark:

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+### Spawning servers
+
+The default Spawner starts servers locally as each user,
+one for each server. These servers listen on localhost,
```
one for each user.

minrk replied:
:heavy_check_mark:

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+
+The Proxy is the public face of the service.
+Users access the server via the proxy.
+By default, this is listening on all public interfaces on port 8000.
```
It is a bit ambiguous as to what the "this is listening" is referring to. Maybe "Users access the server via the proxy, which listens on all public interfaces on port 8000"

minrk replied:
:heavy_check_mark:

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+The Proxy is the public face of the service.
+Users access the server via the proxy.
+By default, this is listening on all public interfaces on port 8000.
+You can access the hub at:
```
Hmmm, now I am confused. Above you introduce 3 main processes (proxy, hub, spawners). Then you describe how the proxy is the main public service that runs on all interfaces on port 8000. But then this sentence calls that "the hub." Do you mean JupyterHub instead of "hub" here, or just call it the "proxy" again to connect it to the above sentences?

minrk replied:
I mean JupyterHub. I don't want to describe it as the proxy, because then people think they want the hub, and skip to using only the hub's private address, which doesn't work (this has actually happened).

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+- `jupyterhub.sqlite` is the sqlite database containing all of the state of the Hub.
+  This file allows the Hub to remember what users are running and where,
+  as well as other information enabling you to 
+  You can change the location of this file with `--db=/path/to/somedb.sqlite`.
```
I would cover the configuration of this below rather than here and just emphasize in this section that a default JupyterHub will auto create these files in the cwd.

minrk replied:
:heavy_check_mark:

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+  You can change the location of this file with `--db=/path/to/somedb.sqlite`.
+- `jupyterhub_cookie_secret` is the encryption key used for securing cookies.
+  This file needs to persist in order for restarting the Hub server to avoid invalidating cookies.
+  Conversely, deleting this file and restarting the server effectively invalidates all login cookies.
```
I would also emphasize that while a default JupyterHub will create both of these files, there location (both) or actual content (just the cookie secret) can be customized as described below.

minrk replied:
:heavy_check_mark:

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+
+## Networking
+
+When it starts, JupyterHub creates two processes:
```
It seems like this section is just repeating the information that is above in the default configuration section. I think it may make sense to combine these two overall section (## default and ## configuring) and then start each sub-section by describing what the default setup is.

minrk replied:
so reorganize the whole thing, so it's:

1. Networking
   1. default
   1. config
1. Security
   1. default
   1. config

etc.?  I think describing the whole default behavior is better, and then a bit of repetition of relevant highlights before diving deeper is better.

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+The proxy is the public-facing part of the application.
+The default public IP is `''`, which means all interfaces on the machine.
+The default port is 8000.
+If you want to specify where the Hub application as a whole can be found,
```
I don't think we should use the name "hub" to refer to both the hub and the whole thing. That is really confusing.

minrk replied:
I agree, but that means massive changes. Would you propose changing the name of the whole application or the hub service?

ellisonbg commented on docs/getting-started.md:
````diff
@@ -0,0 +1,325 @@
+you must set this environment variable before starting the server:
+
+```bash
+export CONFIGPROXY_AUTH_TOKEN=`openssl rand -hex 32`
````
We should mention where this needs to be set (which users env).

ellisonbg commented on docs/getting-started.md:
````diff
@@ -0,0 +1,325 @@
+the value can be loaded from the `JPY_COOKIE_SECRET` env variable:
+
+```bash
+export JPY_COOKIE_SECRET=`openssl rand -hex 1024`
````
Same here

ellisonbg commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+The default PAMAuthenticator is one case of a special kind of authenticator,
+called a LocalAuthenticator,
+indicating that it manages users on the local system.
+When you add a user to the Hub, a LocalAuthenticator checks if that user already exists.
```
Hub->JupyterHub

minrk replied:
Hub is specifically where the user is added.

jdfreder commented:
This is exactly what I need, I'm reviewing this as a true new user! :smile:

jdfreder commented on docs/getting-started.md:
```diff
@@ -0,0 +1,325 @@
+
+The default Spawner starts servers locally as each user,
+one for each server. These servers listen on localhost,
+and start in the given user's home directory.
```
Configurable, right?

jdfreder replied:
Ah I see you mention that this is configurable below in the `Configuring single-user servers` section.  You probably should reference that.

jdfreder commented:
It would be nice to have in this or another doc an outline of the authentication process.  (My interest is running a service using JHub auth, the opposite of using another auth like the GitHub example you include).

minrk commented:
@jdfreder sorry I hadn't actually gotten to the part you are looking for. I think it's an advanced use case that should be separate from this doc.

minrk commented:
Okay, I think I've addressed most of the review. Two bits I haven't dealt with:

- resolving the confusion between Hub (the hub process) and JupyterHub (the whole application)
- restructuring the entire document so that it's broken up by topic, with each little piece of defaults immediately preceding its configuration, getting rid of the complete description of the default behavior.

because I don't quite know how to do either.

@ellisonbg can you provide an outline of how you would prefer the document to be organized? Do you have any proposals for renaming either the hub service in particular, or JupyterHub as a whole? I see the confusion, but both seem to be the right names for the respective elements, unless we switch to non-descriptive names, like "Ganymede".

ellisonbg commented:
I will have a go at editing this and submit a PR to your PR...

minrk commented:
@ellisonbg perfect, thanks.

ellisonbg commented:
OK PR submitted...

On Sun, Mar 29, 2015 at 3:27 PM, Min RK <notifications@github.com> wrote:

> @ellisonbg <https://github.com/ellisonbg> perfect, thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/jupyterhub/pull/191#issuecomment-87490121>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com
